### PR TITLE
Dashboard subvenciones

### LINKED ISSUE #3076: Dashboards / Subvenciones
Mockup: https://www.figma.com/proto/gyMSROblmDqjkKt1TT8BcL5G/Gobierto-%2F-Contratos-Subvenciones?node-id=464%3A0&viewport=607%2C-881%2C0.25&scaling=min-zoom

### PR DESCRIPTION
**Pull request by eltercero**
Closes #3076

## :v: What does this PR do?

Add Subsidies Dashboard with the same features Contracts & tenders Dashboard has.

Note: I want to move some files to another folders, but I'm waiting until the previous PRs are done to do so (to avoid conflicts). When it's ready I'll remove the _blocked_ label. 

## :mag: How should this be manually tested?

https://getafe.gobify.net/dashboards/subvenciones/

## :eyes: Screenshots

Captura de pantalla 2020-05-26 a las 16 16 54

## :shipit: Does this PR changes any configuration file?

No

## :book: Does this PR require updating the documentation?

No

### CODE REVIEW

**Review by ferblape — CHANGES_REQUESTED** [review]
Screenshot 2020-05-26 at 16 49 43

"contratos", copy past is evil 🗡️

File: config/locales/gobierto_dashboards/views/ca.yml [hunk 1]
```diff
@@ -44,9 +44,44 @@ ca:
...
         title: Contractes i licitacions
+      subsidies:
+        amount: Import
+        amount_distribution: Distribució per imports
+        beneficiary: beneficiari
+        category: Categoria
+        date: Data
+        dates: Dates
+        description: Visualització i anàlisi de les subvencions concedides per %{entity_name}
+        fromto: De a Nº Contr.
```
Comment (ferblape) [inline config/locales/gobierto_dashboards/views/ca.yml]: Contr. o Subv?

File: config/locales/gobierto_dashboards/views/en.yml [hunk 2]
```diff
@@ -43,9 +42,43 @@ en:
...
         title: Contracts and tenders
+      subsidies:
+        amount: Amount
+        amount_distribution: Amount distribution
+        beneficiary: Beneficiary
+        category: Category
+        date: Date
+        dates: Dates
+        description: Visualizations and analisys for the granted subsidies from %{entity_name}
+        fromto: From To No Contr.
```
Comment (ferblape) [inline config/locales/gobierto_dashboards/views/en.yml]: Contracts?

File: config/locales/gobierto_dashboards/views/es.yml [hunk 3]
```diff
@@ -43,9 +43,43 @@ es:
...
         title: Contratos y licitaciones
+      subsidies:
+        amount: Importe
+        amount_distribution: Distribución por importes
+        beneficiary: Beneficiario
+        category: Categoría
+        date: Fecha
+        dates: Fechas
+        description: Visualización y análisis de las subvenciones concedidas por %{entity_name}
+        fromto: Desde A Nº. Contr.
```
Comment (ferblape) [inline config/locales/gobierto_dashboards/views/es.yml]: Contr?

**Comment by ferblape** [comment]
Another issue: you are not limiting the top in the summary and it's showing all the assignees

**Comment by eltercero** [comment]
@ferblape fixed translations and table limit. I added it to Contracts too, it wasn't there either 🙊

**Comment by furilo** [comment]
- [ ] Why there are some quantities with the thousand separators and others not? Happens in both dashboards
image

- [x] In Subvenciones/Subvenciones the quantities columns are not right aligned
image

And, fyi, in the Dashboards configuration for the data source with should link to the Dataset, not the Query

**Comment by eltercero** [comment]
>     * [ ]  Why there are some quantities with the thousand separators and others not? Happens in both dashboards

We are using [toLocaleString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString) to handle these (not just in Dashboards) and it seems that this is the way the browser JS engine handles the Spanish locales. It's kind of recent.

Captura de pantalla 2020-05-27 a las 14 31 52

More information: https://stackoverflow.com/questions/57628055/tolocalestring-not-working-on-numbers-less-than-10000-in-all-browsers

We could:

1- Find other external libraries to format these
2- Handle the format ourselves in these cases
3- Use locale `de` when formatting currencies in Spanish (both the easiest and ñapest of them)

**Comment by furilo** [comment]
OK, let's fork the JS locale issue to an issue.

**Comment by eltercero** [comment]
@Crashillo ready to review!

**Comment by eltercero** [comment]
> OK, let's fork the JS locale issue to an issue.

done #3114

File: app/javascript/gobierto_dashboards/index.js [hunk 4]
```diff
@@ -1,16 +1,33 @@
...
+      homeUrl: contractsAppNode.dataset.homeUrl,
+      contractsEndpoint: contractsAppNode.dataset.contractsEndpoint,
+      tendersEndpoint: contractsAppNode.dataset.tendersEndpoint,
+      dataDownloadEndpoint: contractsAppNode.dataset.dataDownloadEndpoint,
+    });
+  }
+});
+
+document.addEventListener('DOMContentLoaded', () => {
+  const subsidiesAppNode = document.getElementById("gobierto-dashboards-subsidies-app");
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/index.js:19]: There's no need to declare the same listener twice

File: app/javascript/gobierto_dashboards/modules/subsidies_controller.js [hunk 5]
```diff
@@ -0,0 +1,317 @@
+import Vue from "vue";
+import VueRouter from "vue-router";
+
+import { sum, mean, median, max } from 'd3-array';
+import { scaleThreshold } from 'd3-scale';
+
+const d3 = { scaleThreshold, sum, mean, median, max }
+
+import * as dc from 'dc'
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/subsidies_controller.js]: Dead code

File: app/javascript/gobierto_dashboards/modules/subsidies_controller.js [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+import crossfilter from 'crossfilter2'
+
+import { getRemoteData, sortByField, uniqueMembers } from '../webapp/lib/utils'
+import { EventBus } from '../webapp/mixins/event_bus'
+
+import { money } from 'lib/shared'
+
+import { AmountDistributionBars } from "lib/visualizations";
+import { GroupPctDistributionBars } from "lib/visualizations";
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/subsidies_controller.js]:
```suggestion
import { AmountDistributionBars, GroupPctDistributionBars } from "lib/visualizations";
```

File: app/javascript/gobierto_dashboards/modules/subsidies_controller.js [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+      domain: [501, 1001, 5001, 10001, 15001],
+      range: [0, 1, 2, 3, 4, 5]
+    };
+    var rangeFormat = d3.scaleThreshold().domain(this._amountRange.domain).range(this._amountRange.range);
+
+    for(let i = 0; i < subsidiesData.length; i++){
+      const subsidy = subsidiesData[i];
+      const amount = subsidy.amount ? parseFloat(subsidy.amount) : 0.0
+
+      let [beneficiary_id, ...beneficiary_name] = subsidy.beneficiary.split(' ')
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/subsidies_controller.js:125]: Magic is coming! 👏

File: app/javascript/gobierto_dashboards/modules/subsidies_controller.js [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+    this.charts['dates'] = new GroupPctDistributionBars(renderOptions);
+  }
+
+  _updateChartsFromFilter(options){
+    const container = this.charts[options.id].container;
+
+    if (options.all) {
+      container.filter(null);
+      container.filter([options.titles]);
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/subsidies_controller.js:296]: Explain why is that

File: app/javascript/gobierto_dashboards/modules/subsidies_controller.js [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    });
+  }
+
+  _currentDataSource(){
+    return this.reduced || this.data
+  }
+
+  _formalizedSubsidiesData(subsidiesData){
+    return subsidiesData
+  }
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/modules/subsidies_controller.js]: I wonder what this do....

File: app/javascript/gobierto_dashboards/webapp/components/DownloadButton.vue [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      default: ''
+    }
+  },
+  created(){
+    this.isShowDownloadDatalinkShown = !!this.dataDownloadEndpoint
+  },
+  methods: {
+    showDownloadDataLink(){
+      return !!this.dataDownloadEndpoint
+    }
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/components/DownloadButton.vue]: When is this called?

File: app/javascript/gobierto_dashboards/webapp/containers/subsidies_shared/Nav.vue [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    <ul>
+      <router-link
+        :to="{ name: 'summary' }"
+        :class="{ 'is-active': activeTab === 0 }"
+        tag="li"
+        class="dashboards-home-nav--tab"
+        @click.native="markTabAsActive(0)"
+      >
+        <i class="fas fa-chart-bar" />
+        <i class="far fa-chart-bar" />
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/subsidies_shared/Nav.vue:12]: I've seen this. @jorgeatgu, do you remember what happened if we remove the duplicated?

File: app/javascript/gobierto_dashboards/webapp/containers/subsidy/SubsidiesIndex.vue [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    Table
+  },
+  data() {
+    return {
+      subsidiesData: this.$root.$data.subsidiesData,
+      items: []
+    }
+  },
+  created() {
+    EventBus.$on('refresh_summary_data', () => {
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/subsidy/SubsidiesIndex.vue:27]: $on but no $off

File: app/javascript/gobierto_dashboards/webapp/containers/subsidies_shared/Aside.vue [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    dataDownloadEndpoint: {
+      type: String,
+      default: ''
+    }
+  },
+  created(){
+    this.initFilterOptions();
+    this.updateCounters(true);
+
+    EventBus.$on('dc_filter_selected', ({title, id}) => {
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/containers/subsidies_shared/Aside.vue:74]: $on but no $off

File: app/javascript/lib/shared/modules/vue-filters.js [hunk 14]
```diff
@@ -27,10 +27,15 @@ export const date = (value, opts = {}) => {
 }
 
 export const truncate = (value, opts = {}) => {
-  const omission = opts['omission'] || '...';
   const length = opts['length'] || 30;
+  const str = `${value || ''}`
```
Comment (Crashillo) [inline app/javascript/lib/shared/modules/vue-filters.js:31]: Iterating over your commons. That's the way! 🥳

File: app/views/gobierto_dashboards/dashboards/subsidies.html.erb [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+      <h2 class="with_description">
+        <%= t(".title") %>
+      </h2>
+      <p>
+        <%= t(".description", entity_name: current_site.name) %>
+      </p>
+    </div>
+  </div>
+
+  <div class="js-loading">
```
Comment (Crashillo) [inline app/views/gobierto_dashboards/dashboards/subsidies.html.erb:15]: I think you can add this stuff in the vue app

File: config/locales/gobierto_dashboards/views/ca.yml [hunk 16]
```diff
@@ -44,11 +43,45 @@ ca:
...
+        subsidies: Subvencions
+        summary:
+          collective_subsidies_for: a col·lectius per import de
+          individual_subsidies_for: a particulars per import de
+          label_half_spendings_subsidies: El <strong> <span id = "half-spendings-subsidies-pct">
+            </ span> </ strong> de subvencions concentren el 50% de tota la despesa
+          label_larger_subsidy_amount: La major subvenció suposa <strong> <span id
+            = "larger-subsidy-amount-pct"> </ span> </ strong> de tota la despesa
+            en subvencions
+          label_less_than_1000: El <strong> <span id = "less-than-1000-pct"> </ span>
```
Comment (Crashillo) [inline config/locales/gobierto_dashboards/views/ca.yml]: Back again?

File: config/locales/gobierto_dashboards/views/ca.yml [hunk 17]
```diff
@@ -44,11 +43,45 @@ ca:
...
+            </ strong> de les subvencions són menors de <strong> 1.000 € </ strong>
+          mean_amount: Import mitjà
+          median_amount: Import mitjà
```
Comment (Crashillo) [inline config/locales/gobierto_dashboards/views/ca.yml]: You cannot use the same label for different things, median (maths) is mediano/a

**Review by Crashillo — CHANGES_REQUESTED** [review]

File: app/javascript/gobierto_dashboards/index.js (hunk 4 above)
Comment (eltercero) [reply app/javascript/gobierto_dashboards/index.js:19]: 🤦

File: app/javascript/gobierto_dashboards/webapp/components/DownloadButton.vue (hunk 10 above)
Comment (eltercero) [reply app/javascript/gobierto_dashboards/webapp/components/DownloadButton.vue]: It's not, sorry! It was done between reviews and forgot to remove that one.

File: config/locales/gobierto_dashboards/views/ca.yml (hunk 17 above)
Comment (eltercero) [reply config/locales/gobierto_dashboards/views/ca.yml]: I know mean and median are different, but don't they use the same word in catalan? If not, which is it?

Comment (ferblape) [reply config/locales/gobierto_dashboards/views/ca.yml]: Yes, they use the same words :D but there are synomis. You can use mitja for mean and mitjana for median

Comment (eltercero) [reply config/locales/gobierto_dashboards/views/ca.yml]: 👍 I'll change it

File: app/javascript/lib/shared/modules/vue-filters.js (hunk 14 above)
Comment (eltercero) [reply app/javascript/lib/shared/modules/vue-filters.js:31]: I don't understand what you meant, is it ok then 😅 ?

File: config/locales/gobierto_dashboards/views/ca.yml (hunk 16 above)
Comment (eltercero) [reply config/locales/gobierto_dashboards/views/ca.yml]: Still not sure about the best way to handle these, tbh. Separation is important, but it'll also be messier to handle.

File: config/locales/gobierto_dashboards/views/ca.yml (hunk 17 above)
Comment (Crashillo) [reply config/locales/gobierto_dashboards/views/ca.yml]: https://ca.wikipedia.org/wiki/Mediana

File: app/javascript/lib/shared/modules/vue-filters.js (hunk 14 above)
Comment (Crashillo) [reply app/javascript/lib/shared/modules/vue-filters.js:31]: Yes, it's a compliment. :)

File: config/locales/gobierto_dashboards/views/ca.yml (hunk 16 above)
Comment (Crashillo) [reply config/locales/gobierto_dashboards/views/ca.yml]: My point of view here is, let's imagine we move the translations to a professional tool, like transifex, why a translator have to know there are HTML in the text and not only this, but even the HTML contains required identificators for the app?

If the translation was: `label_larger_subsidy_amount: La <h1>major</h1> subvenció...`, it wouldn't be OK, but.... mmmm reasonable. But including identificadors? no way.

On the other hand, someone else is developing the controller and need to debug the values. Where are they? OMG, embedded in the translation texts. This is about [comparing apples and oranges](https://www.wordreference.com/es/en/translation.asp?spen=mezclar%20las%20churras%20con%20las%20merinas). Although split the sentence wouldn't be neither the best approach, at least the developer knows where its code is.

Comment (eltercero) [reply config/locales/gobierto_dashboards/views/ca.yml]: I did the split, you may check now 👍

File: app/javascript/gobierto_dashboards/webapp/components/DownloadButton.vue [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+  },
+  props: {
+    dataDownloadEndpoint: {
+      type: String,
+      default: ''
+    }
+  },
+  created(){
+    this.isShowDownloadDatalinkShown = !!this.dataDownloadEndpoint
```
Comment (Crashillo) [inline app/javascript/gobierto_dashboards/webapp/components/DownloadButton.vue]: When you declare a property with `this`, and then you use it in the DOM, it _should_ be declared in the `data` function.
However, in this case, you can save declaring a new variable for that and simply as:
`v-if="!!this.dataDownloadEndpoint"`

File: config/locales/gobierto_dashboards/views/en.yml [hunk 19]
```diff
@@ -30,24 +29,57 @@ en:
         summary:
           contracts: Assigned contracts
           contracts_for: contracts for a total amount of
-          label_half_spendings_contracts: <strong><span id="half-spendings-contracts-pct"></span></strong>
-            of contracts accumulate the 50% of all the spending
-          label_larger_contract_amount: The largest contract means a <strong><span
-            id="larger-contract-amount-pct"></span></strong> of all the spending
-          label_less_than_1000: <strong><span id="less-than-1000-pct"></span></strong>
-            of contracts are less than <strong>1.000 €</strong>
+          label_half_spendings_contracts_1: ""
```
Comment (Crashillo) [inline config/locales/gobierto_dashboards/views/en.yml:32]: Could be another option declare variables? For instance:

yml: `%{value} of contracts accumulate...`
vue: `<p class="description" v-html="I18n.t(LABEL, { value: <strong>VARIABLE</strong> } )">`

Just asking

**Review by Crashillo — APPROVED** [review]
You can implement the suggestions in the next PR

File: config/locales/gobierto_dashboards/views/en.yml (hunk 19 above)
Comment (eltercero) [reply config/locales/gobierto_dashboards/views/en.yml:32]: I tried that and it didn't work. Maybe I did something wrong.

File: app/javascript/gobierto_dashboards/webapp/containers/subsidies_shared/Nav.vue (hunk 11 above)
Comment (jorgeatgu) [reply app/javascript/gobierto_dashboards/webapp/containers/subsidies_shared/Nav.vue:12]: Nothing happens, is the same icon but with different weight

far class = font-awesome Regular
fas class = font-awesome Solid

**Pull request merged**